### PR TITLE
Enable to provide a custom logger in the API

### PR DESCRIPTION
**Pull request by TheMaskedTurtle**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
The purpose is to able to call the dichotomy with the client logger implementation. Based on slf4j-api.

### CODE REVIEW

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java [hunk 1]
```diff
@@ -78,12 +87,12 @@ public DichotomyEngine(Index<T> index, IndexStrategy indexStrategy, NetworkShift
         network.getVariantManager().cloneVariant(initialVariant, newVariant);
         network.getVariantManager().setWorkingVariant(newVariant);
         try {
-            LOGGER.debug("Shifting network");
-            networkShifter.shiftNetwork(stepValue, network);
-            LOGGER.debug("Validating network");
-            return networkValidator.validateNetwork(network);
+            logger.debug("Shifting network");
```
Comment (benrejebmoh) [inline farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java:90]: add more context: network id, step value

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java [hunk 2]
```diff
@@ -78,12 +87,12 @@ public DichotomyEngine(Index<T> index, IndexStrategy indexStrategy, NetworkShift
...
+            networkShifter.shiftNetwork(stepValue, network, logger);
+            logger.debug("Validating network");
```
Comment (benrejebmoh) [inline farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java:92]: same

File: farao-dichotomy-api/src/test/java/com/farao_community/farao/dichotomy/api/NetworkValidatorMock.java [hunk 3]
```diff
@@ -21,7 +21,12 @@ public NetworkValidatorMock(double limit) {
     }
 
     @Override
-    public DichotomyStepResult<Object> validateNetwork(Network network) throws ValidationException {
+    public DichotomyStepResult<Object> validateNetwork(Network network) {
```
Comment (benrejebmoh) [inline farao-dichotomy-api/src/test/java/com/farao_community/farao/dichotomy/api/NetworkValidatorMock.java:24]: i would remove or deprecate it. I think breaking compatibility, still better than breaking behaviour..

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java [hunk 4]
```diff
@@ -33,19 +33,27 @@
...
     private final Index<T> index;
     private final IndexStrategy indexStrategy;
     private final NetworkShifter networkShifter;
     private final NetworkValidator<T> networkValidator;
     private final int maxIteration;
+    private final Logger logger;
 
     public DichotomyEngine(Index<T> index, IndexStrategy indexStrategy, NetworkShifter networkShifter, NetworkValidator<T> networkValidator) {
-        this(index, indexStrategy, networkShifter, networkValidator, DEFAULT_MAX_ITERATION_NUMBER);
+        this(index, indexStrategy, networkShifter, networkValidator, DEFAULT_MAX_ITERATION_NUMBER, LoggerFactory.getLogger(DichotomyEngine.class));
```
Comment (benrejebmoh) [inline farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java:45]: I m not convinced by the hack. Mixing technical params (logger ) with business ones . Here you are defining the logger as a key param for the process behind,.. which is not.
I would create in a subLevel a pubic class dedicated for business logging, then use it to separate business from techical logs.

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java [hunk 5]
```diff
@@ -54,21 +62,22 @@ public DichotomyEngine(Index<T> index, IndexStrategy indexStrategy, NetworkShift
...
-            LOGGER.info("Validating step value '{}'", nextValue);
+            logger.info("Validating step value '{}'", nextValue);
             DichotomyStepResult<T> dichotomyStepResult = validate(nextValue, network, initialVariant);
             index.addDichotomyStepResult(nextValue, dichotomyStepResult);
             iterationCounter++;
         }
 
         if (iterationCounter == maxIteration) {
-            LOGGER.error("Max number of iteration reached during dichotomy, research precision has not been reached.");
+            logger.error("Max number of iteration reached during dichotomy, research precision has not been reached.");
```
Comment (benrejebmoh) [inline farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java:80]: remind the maxIteration in the log

**Review by benrejebmoh — CHANGES_REQUESTED** [review]

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/logging/DichotomyLoggerProvider.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.farao_community.farao.dichotomy.api.logging;
+
+import java.text.DecimalFormat;
+
+/**
+ * @author Peter Mitri {@literal <peter.mitri at rte-france.com>}
```
Comment (benrejebmoh) [inline farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/logging/DichotomyLoggerProvider.java:13]: author name is not correct

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java [hunk 7]
```diff
@@ -61,14 +60,19 @@ public DichotomyEngine(Index<T> index, IndexStrategy indexStrategy, NetworkShift
         String initialVariant = network.getVariantManager().getWorkingVariantId();
         while (!index.precisionReached() && iterationCounter < maxIteration) {
             double nextValue = indexStrategy.nextValue(index);
-            LOGGER.info("Validating step value '{}'", nextValue);
+            BUSINESS_LOGS.info("Next dichotomy step: {}", DECIMAL_FORMAT.format(nextValue));
             DichotomyStepResult<T> dichotomyStepResult = validate(nextValue, network, initialVariant);
+            if (dichotomyStepResult.isValid()) {
+                BUSINESS_LOGS.info("Network at dichotomy step {} is secure", DECIMAL_FORMAT.format(nextValue));
```
Comment (benrejebmoh) [inline farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java:66]: i would use directly String.format("%.2f", nextValue)

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/logging/DichotomyLoggerProvider.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ */
+public final class DichotomyLoggerProvider {
+    public static final DichotomyLogger BUSINESS_LOGS = new DichotomyBusinessLogs();
+    public static final DichotomyLogger BUSINESS_WARNS = new DichotomyBusinessWarns();
+    public static final DichotomyLogger TECHNICAL_LOGS = new TechnicalLogs();
+    public static final DecimalFormat DECIMAL_FORMAT = new DecimalFormat("#.##");
```
Comment (benrejebmoh) [inline farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/logging/DichotomyLoggerProvider.java:19]: DECIMAL_FORMAT doesn't fit well here

**Review by benrejebmoh — CHANGES_REQUESTED** [review]

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/logging/DichotomyLoggerProvider.java (hunk 6 above)
Comment (TheMaskedTurtle) [reply farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/logging/DichotomyLoggerProvider.java:13]: changed

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java (hunk 7 above)
Comment (TheMaskedTurtle) [reply farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java:66]: done

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/logging/DichotomyLoggerProvider.java (hunk 8 above)
Comment (TheMaskedTurtle) [reply farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/logging/DichotomyLoggerProvider.java:19]: removed

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java (hunk 5 above)
Comment (TheMaskedTurtle) [reply farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java]: done

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java (hunk 4 above)
Comment (TheMaskedTurtle) [reply farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java]: changed

File: farao-dichotomy-api/src/test/java/com/farao_community/farao/dichotomy/api/NetworkValidatorMock.java (hunk 3 above)
Comment (TheMaskedTurtle) [reply farao-dichotomy-api/src/test/java/com/farao_community/farao/dichotomy/api/NetworkValidatorMock.java:23]: obsolete

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java (hunk 2 above)
Comment (TheMaskedTurtle) [reply farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java]: removed

File: farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java (hunk 1 above)
Comment (TheMaskedTurtle) [reply farao-dichotomy-api/src/main/java/com/farao_community/farao/dichotomy/api/DichotomyEngine.java]: removed

**Pull request merged**